modules/effects/src/effects_module.ts

```diff
@@ -46,16 +64,44 @@ export class EffectsModule {
         EffectSources,
         Actions,
         rootEffects,
+        {
+          provide: USER_PROVIDED_ROOT_EFFECTS,
+          multi: true,
+          useValue: [],
+        },
+        {
+          provide: _ROOT_EFFECTS,
+          useValue: rootEffects,
+        },
         {
           provide: ROOT_EFFECTS,
-          deps: rootEffects,
-          useFactory: createSourceInstances,
+          useFactory: createEffects,
+          deps: [Injector, _ROOT_EFFECTS, USER_PROVIDED_ROOT_EFFECTS],
         },
       ],
     };
   }
 }
 
+export function createEffects(
+  injector: Injector,
+  effects: Type<any>[],
+  userProvidedEffectGroups: Type<any>[][]
+): any[] {
+  const mergedEffects: Type<any>[] = effects;
```

Comment:
How about the following
```ts
const mergedEffects: Set<Type<any>> = new Set([...userProvidedEffectGroups.flat(), ...effects]);
```
this is all remove the duplicates, if there are any.

Reply:
I think the 'flat' method is not yet supported by typescript (So it tells you something like this)
![Screenshot_2](https://user-images.githubusercontent.com/15183853/72974827-084d5880-3dd0-11ea-9068-448dec9fbf23.png)

That's also the reason i had to spread so much

Reply:
I see... `flat` is not supported on edge, so I many we should avoid it as well for now.

Still, let's not use `forEach` of the Array, let's do it with `for ... of` loop.

Reply:
Spreads are fine :)

As for the `forEach`, here is the explanation why not to use it:
> Do not use Array.prototype.forEach, it makes code harder to debug and defeats some useful compiler checks (e.g. reachability).

Reply:
Used a `for ... of` loop 👍